src/psbt.c

```diff
@@ -0,0 +1,2302 @@
+            for (i = 0; i < keypaths->num_items; ++i) {
+                if (keypaths->items[i].origin.path) {
+                    wally_clear(keypaths->items[i].origin.path, keypaths->items[i].origin.path_len);
+                }
+            }
+            clear_and_free(keypaths->items, keypaths->num_items * sizeof(*keypaths->items));
+        }
+        wally_clear(keypaths, sizeof(*keypaths));
+        wally_free(keypaths);
+    }
+    return WALLY_OK;
+}
+
+static struct keypath_map *clone_keypath_map(const struct keypath_map *keypaths)
+{
+    struct keypath_map *result;
+    size_t i;
+    int ret;
+
+    ret = keypath_map_init_alloc(keypaths->items_allocation_len, &result);
+
+    if (ret == WALLY_OK) {
+        for (i = 0; i < keypaths->num_items && ret == WALLY_OK; ++i) {
+            if (!memcpy(&result->items[i].pubkey, keypaths->items[i].pubkey, 65) ||
+                !memcpy(&result->items[i].origin.fingerprint, keypaths->items[i].origin.fingerprint, 4)) {
+                ret = WALLY_ENOMEM;
+            }
+            if (ret == WALLY_OK && keypaths->items[i].origin.path) {
+                if (!clone_bytes((unsigned char **)&result->items[i].origin.path, (unsigned char *)keypaths->items[i].origin.path, keypaths->items[i].origin.path_len * sizeof(*keypaths->items[i].origin.path))) {
+                    ret = WALLY_ENOMEM;
+                }
+                result->items[i].origin.path_len = keypaths->items[i].origin.path_len;
+            }
+
+            if (ret != WALLY_OK) {
+                keypath_map_free(result);
+            }
+        }
+        result->num_items = keypaths->num_items;
+    }
+    return ret == WALLY_OK ? result : NULL;
+}
+
+int add_new_keypath(struct keypath_map *keypaths,
+    unsigned char pubkey[65],
+    unsigned char fingerprint[4],
+    uint32_t *path,
+    size_t path_len)
+{
+    size_t latest;
+
+    if (keypaths->num_items == keypaths->items_allocation_len) {
+        if (keypaths->items_allocation_len == 0) {
+            keypaths->items_allocation_len = 1;
+        }
+        keypaths->items_allocation_len *= 2;
+        struct keypath_map *new_keypaths = clone_keypath_map(keypaths);
+        if (!new_keypaths) {
+            return WALLY_ENOMEM;
+        }
+        keypaths->items = new_keypaths->items;
+        keypaths->num_items = new_keypaths->num_items;
+        keypaths->items_allocation_len = new_keypaths->items_allocation_len;
+        wally_free(new_keypaths);
+    }
+
+    latest = keypaths->num_items;
+
+    if (!memcpy(&keypaths->items[latest].pubkey, pubkey, 65) ||
+        !memcpy(&keypaths->items[latest].origin.fingerprint, fingerprint, 4)) {
+        return WALLY_ENOMEM;
+    }
+    if (path) {
+        if (!clone_bytes((unsigned char **)&keypaths->items[latest].origin.path, (unsigned char *)path, path_len * sizeof(*path))) {
+            return WALLY_ENOMEM;
+        }
+        keypaths->items[latest].origin.path_len = path_len;
+    }
+    keypaths->num_items++;
```

Comment:
it'd appear if path is null it will still increase num_items so it seems to assume null paths can be added. is this the intention?

Reply:
That is intended.